modules/ctransformers_model.py

```diff
@@ -9,29 +12,57 @@ class CtransformersModel:
     def __init__(self):
         pass
 
-    @classmethod
     def from_pretrained(self, path):
-        result = self()
 
         config = AutoConfig.from_pretrained(
             str(path),
             threads=shared.args.threads,
             gpu_layers=shared.args.n_gpu_layers,
             batch_size=shared.args.n_batch,
+            context_length=shared.args.n_ctx,
             stream=True
         )
 
+        model_name_dir = self.model_dir(path)
         self.model = AutoModelForCausalLM.from_pretrained(
-            str(result.model_dir(path) if result.model_type_is_auto() else path),
-            model_type=(None if result.model_type_is_auto() else shared.args.model_type),
+            str(model_name_dir if self.model_type_is_auto() else path),
+            model_type=(self.model_try_infer_type(model_name_dir, shared.args.model_type)),
             config=config
         )
 
         logger.info(f'Using ctransformers model_type: {self.model.model_type} for {self.model.model_path}')
-        return result, result
+        return self, self
 
     def model_type_is_auto(self):
-        return shared.args.model_type == "Auto" or shared.args.model_type == "None"
+        return shared.args.model_type is None or shared.args.model_type == "Auto" or shared.args.model_type == "None"
+
+    # llama is more of a fallback option
+    match_table = OrderedDict({
+        'falcon': re.compile('.*falcon.*ggml.*\\.bin'),
+        'mpt': re.compile('.*mpt.*ggml.*\\.bin'),
+        'starcoder': re.compile('.*star(chat|coder).*ggml.*\\.bin'),
+        'llama': re.compile('.*ggml.*\\.bin')
+    })
```

Comment:
The webui has its own mechanism for inferring and setting parameters based on model names. The patterns are at `models/config.yaml`, and there you can set `model_type` based on regexes. So this should be removed and moved there.

Reply:
I have dropped this part in favor of the more conservative `models/config.yaml` changes.